File: app/src/main/java/com/example/installer/adapter/BottomSheetRecyclerAdapter.java
package com.example.installer.adapter;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

/**
 * Author:  ZhangTao
 * Date: 2018/3/5.
 */

public abstract class BottomSheetRecyclerAdapter<T> extends RecyclerView.Adapter<CommonViewHolder> {
    private List<T> dataSource;

    public BottomSheetRecyclerAdapter(List<T> dataSource) {
        this.dataSource = dataSource;
    }

    @Override
    public CommonViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View itemView = LayoutInflater.from(parent.getContext()).inflate(getItemLayoutID(), parent, false);
        return new CommonViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(CommonViewHolder holder, int position) {
        convert(holder, dataSource.get(position), position);
    }


    public abstract void convert(CommonViewHolder holder, T item, int position);

    public abstract int getItemLayoutID();

    @Override
    public int getItemCount() {
        return dataSource == null ? 0 : dataSource.size();
    }

}

File: app/src/main/java/com/example/installer/adapter/BottomSheetRecyclerAdapter.kt
package com.example.installer.adapter

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.annotation.NonNull
import androidx.recyclerview.widget.RecyclerView

/**
@author zhangtao
@date   2020/11/3

 **/
abstract class BottomSheetRecyclerAdapter<T>(private val dataSource: List<T>) :
    RecyclerView.Adapter<CommonViewHolder>() {

    override fun onCreateViewHolder(@NonNull parent: ViewGroup, viewType: Int): CommonViewHolder {
        val itemView = LayoutInflater.from(parent.context).inflate(getItemLayoutID(), parent, false)
        return CommonViewHolder(itemView)
    }

    override fun onBindViewHolder(@NonNull holder: CommonViewHolder, position: Int) {
        convert(holder, dataSource[position], position)
    }

    override fun getItemCount(): Int {
        return dataSource.size
    }

    abstract fun convert(holder: CommonViewHolder, item: T, position: Int)

    abstract fun getItemLayoutID(): Int
}